tsdfileapi/api.py

```diff
@@ -738,27 +747,53 @@ def prepare(self):
                 logging.error(e)
                 raise Exception
             try:
+                # 1. Check tenant reference
                 try:
                     tenant = tenant_from_url(self.request.uri)
                     self.tenant = tenant
                     assert options.valid_tenant.match(tenant)
                 except AssertionError as e:
                     logging.error('URI does not contain a valid tenant')
                     raise e
+                # 2. get group param
                 try:
                     self.group_name = url_unescape(self.get_query_argument('group'))
                 except Exception:
                     self.group_name = tenant + '-member-group'
                 filemode = filemodes[self.request.method]
+                # 3. start processing the data
                 try:
-                    # optionally create the tenant_dir
-                    if options.create_tenant_dir:
-                        if not os.path.lexists(self.tenant_dir):
-                            os.makedirs(self.tenant_dir)
+                    # 3.1 extract info from uri
                     content_type = self.request.headers['Content-Type']
                     uri_filename = self.request.uri.split('?')[0].split('/')[-1]
                     filename = check_filename(url_unescape(uri_filename),
                                               disallowed_start_chars=options.start_chars)
+                    # 3.2 optionally create dirs
+                    # 3.2.1 tenant dir
+                    if options.create_tenant_dir:
+                        if not os.path.lexists(self.tenant_dir):
+                            os.makedirs(self.tenant_dir)
+                    # 3.2.2 destination dir
+                    self.resource_dir = None
+                    try:
+                        resource_references = self.request.uri.split('?')[0].split('/')[5:]
+                        url_dirs = url_unescape('/'.join(resource_references[:-1]))
+                        self.resource_dir = os.path.normpath(f'{self.tenant_dir}/{url_dirs}')
+                        if not os.path.lexists(self.resource_dir):
+                            logging.info(f'creating resource dir: {self.resource_dir}')
+                            os.makedirs(self.resource_dir)
+                            target = self.tenant_dir
+                            for _dir in url_dirs.split('/'):
+                                target += f'/{_dir}'
+                                try:
+                                    if self.group_config['enabled']:
+                                        subprocess.call(['chmod', '2770', target])
```

Comment:
Perhaps this subprocess call could be replaced with os.chmod() if privilege elevation isn't needed?

Reply:
Sure thing.